### DISCUSSION TITLE
Document Structured Concurrency behaviour

### DISCUSSION THREAD

**Discussion by alexito4** [pointfreeco/swift-composable-architecture#1481] [General]
Hello!
I was migrating to the new async primitives and found myself wondering if they do follow structured concurrency properly. For `.task` and `.run` that seemed quite obvious, but I had my doubts about `.fireAndForget`.

`.fireAndForget` description and name made me think it might have been tailored for similar usages as `Task.detached`. I suspected that was not the case, since one barely needs that, but the documentation didn't clarify it for me. A quick read trough the source code showed me that all those primitives follow the same implementation and thus follow structured concurrency (❤️ ) but I run some tests to convince myself.

I was thinking if it's worth adding a note on ?at least? `.fireAndForget` to clarify that it will be cancelled properly, even if we don't care about its result. It would also help my fear of this being not set in stone and changing in a future version 😆 

_I'm happy making a PR with a suggestion, just wanted to know if you want to talk about this in the docs or is too much to explain._

**Comment by mbrandonw** [comment]
Hi @alexito4, you are correct that `fireAndForget` does not use `Task.detached` under the hood, and instead uses a regular `Task`. As you mention, there's really very few reasons to reach for `Task.detached`, and in particular now that the entire dependency system of TCA runs off of task locals, detaching comes with the additional "quirk" that dependencies get reset back to their defaults.

We are definitely open to clarifying some docs for the method. Feel free to open a PR with your ideas, and then we can give some feedback.

**Comment by tgrapperon** [comment]
@alexito4 Indeed, I already stumbled upon this in the past. I'm wondering if "fire and forget" is perceived differently for native speakers. There is a good chance the the upcoming presentation & navigation APIs will automatically cancel presented/navigated effects when dismissing/popping. This includes `.fireAndForget` ones, which I find slightly counter-intuitive looking at it from the presented feature's point of view. This will probably be discussed at this point, but I feel that `.fireAndForget` should be allowed to go through, but I don't know what are the consequences down the line, as there are also uses for a `.fireAndForget` that cancels.

**Reply by mbrandonw** [reply]
It's definitely an interesting thing to consider that I haven't thought much about. If I think really intensely about it, I can convince myself that "fire-and-forget" means to truly escape the boundaries of the library, and so shouldn't participate in cancellation. But, at the same time, I think the majority of people will want cancellation, and if it does not cancel people will encounter strange bugs. And on top of that, I think 100% of people will want the current dependencies to be passed to the effect.

If you truly want to escape cancellation you can always do this:

```swift
return .fireAndForget {
  Task { // or Task.detached
  }
}
```

The funny thing is that technically all of `Effect.task`/`.run`/`.fireAndForget` are powered by the same thing, which is `Effect.run`. We could just drop `.fireAndForget` and if you want an effect that doesn't emit anything you just do:

```swift
return .run { _ in 
  try await self.analytics.trackEvent("Button Tapped")
}
```

The `_ in` would make it clear you don't send anything back into the system. Or maybe there could even be an overload of `.run` that doesn't take an argument, though not sure if that would cause ambiguity issues eventually.

And if we go that route I start to wonder if even `Effect.task` pulls its weight. Maybe everything should just be `Effect.run`. 🤔

These are just idle thoughts. I haven't talked to Stephen about any of this yet, though we did certainly consider this approach when we originally worked on the concurrency tools many months ago.

**Reply by davdroman** [reply]
> And if we go that route I start to wonder if even `Effect.task` pulls its weight. Maybe everything should just be `Effect.run`. 🤔

This is definitely something that feels a bit jarring when using the new concurrency tools.

I always have to remind myself which function does what, whereas if we just had `Effect.run` for everything it could potentially simplify that mental model without much compromise in ergonomics.

**Reply by tgrapperon** [reply]
@mbrandonw We'll probably have the opportunity to discuss the topic during the navigation beta. I think the problem is more with my perception of `.fireAndForget` which I imagine like dropping a letter into the mailbox. In my case, I had a shinny async/await CoreData wrapper, and in a presented feature I was sending a `.fireAndForget { try? await self.database.save() }`. But automatic cancellation on dismissal (or at the time, a small bug now fixed that canceled the effect very shortly after) was able to fish out the `save` and cancel it in flight, whereas in my mind, all ties were cut and it was a done deal. 

At the end of the day, it is probably a design problem on my side (I should probably implement resilient end-points), and I can fortunately use a `Task` to escape the effect's context [^1]. I think that the current behavior is fine, and everything should cancel by default. I'm simply wondering what I should change in my mind to be able to think about this issue without having to experience the bug. I'm sure that if you do a poll, you will find a non-negligible fraction of users assuming that a `.fireAndForget` is not cancellable. Maybe a special API on `.fireAndForget` that would do that could help with discoverability, albeit it's already quite confusing when you mix with Combine ones.

We should probably design our features defensively locally, and explicitly make sure at this level that nothing can cancel it if we intend one effect to conclude. In other words, one must acknowledge locally the existence of "the dark forest", that is the existence of higher-level entities capable of "killing" your effects[^2]. This is not something that I was doing, as I was usually designing features as their own closed universe. I think that all my "troubles" arise from this point of view that I need to evolve.

I think that the current `.task/.run/.fireAndForget` is fine. It is very nice that everything is ultimately equivalent to a `.run`, and new patterns may emerge, but the `.task` and `.fireAndForget` specializations are very nice to have. One can also imagine reducers evolving up to a point where everything could be able to be expressed inline into a `reduce(state, action) async -> Void`, with a `self.send(…)` provided by the reducer. 

[^1]: Because of your "concurrency" videos series, I'm overly cautious when creating new `Task`s, up to a point where I often won't consider it even when the unpropagated cancellation is indeed exactly what's needed!

[^2]: As I'm thinking more about it, this "aether" or "common playground" is the current async context, which is something new since the concurrency update.
